File: agent/query_engine/parse_sse_to_engine_event.py
from __future__ import annotations

import json
from typing import Any

from query_engine.EngineEvent import EngineEvent


def parse_sse_to_engine_event(raw: str) -> EngineEvent:
    event_name = "message"
    data: dict[str, Any] = {}
    for line in raw.strip().split("\n"):
        if line.startswith("event:"):
            event_name = line.split(":", 1)[1].strip()
        elif line.startswith("data:"):
            payload = line.split(":", 1)[1].strip()
            try:
                parsed = json.loads(payload)
                if isinstance(parsed, dict):
                    data = parsed
            except json.JSONDecodeError:
                data = {}

    payload = data.get("payload") if isinstance(data.get("payload"), dict) else data
    source = str(data.get("source", "system"))
    trace_id = str(data.get("trace_id", "")) if data.get("trace_id") is not None else None
    event_version = str(data.get("event_version", "1.0"))

    return EngineEvent(
        event=event_name,
        source=source,
        payload=payload if isinstance(payload, dict) else {},
        trace_id=trace_id,
        event_version=event_version,
    )

File: agent-ts/src/query_engine/parse_sse_to_engine_event.ts
import type { JsonObject } from "../common/json/types/JsonTypes.js";
import { EngineEvent } from "./EngineEvent.js";

export function parseSseToEngineEvent(rawEvent: string): EngineEvent {
  const lines = rawEvent.replace(/\r/g, "").split("\n");
  let eventName = "";
  let dataLine = "";

  for (const line of lines) {
    if (line.startsWith("event:")) {
      eventName = line.slice(6).trim();
    } else if (line.startsWith("data:")) {
      dataLine = line.slice(5).trim();
    }
  }

  if (!eventName) {
    throw new Error("invalid sse event: missing event name");
  }
  if (!dataLine) {
    throw new Error("invalid sse event: missing data");
  }

  let parsed: unknown;
  try {
    parsed = JSON.parse(dataLine);
  } catch {
    throw new Error("invalid sse event: data is not json");
  }

  const payload = typeof parsed === "object" && parsed !== null ? parsed : {};
  const record = payload as Record<string, unknown>;
  return new EngineEvent(
    eventName,
    typeof record.source === "string" ? record.source : "system",
    typeof record.payload === "object" && record.payload !== null ? (record.payload as JsonObject) : {},
    typeof record.traceId === "string" ? record.traceId : null,
    typeof record.eventVersion === "string" ? record.eventVersion : "1.0"
  );
}
